pylearn2/datasets/dataset.py

```diff
@@ -48,6 +48,28 @@ def iterator(self, mode=None, batch_size=None, num_batches=None,
             through the dataset and may potentially be shared by
             multiple iterator objects simultaneously (see "Notes"
             below).
+        data_specs : (space, source) pair, optional
+            `space` is an instance of `Space` (possibly a
+            `CompositeSpace`), and `source` is a string (or tuple of
+            strings if `space` is a `CompositeSpace`). `source` defines
+            which data should be returned by the iterator (for instance
+            'features', 'targets'), and `space` specifies in which
+            format it will be returned. These specifications should be
+            flat: if `space` is a `CompositeSpace`, the sub-spaces should
+            not be `CompositeSpace`s, and `source` should not be a nested
+            tuple, and the same sub-source should not be mentioned more
+            than once with the same sub-space. If `data_specs` is not
```

Comment:
It is not that all sources have to be unique, it is that all elementary (source, space) must be unique.
For instance, if two different costs want the same source (`'targets'`) in different formats (one-hot vs. index), in addition to the features, then the source strings are not unique: `'targets'` will be specified twice, once in conjunction with a `VectorSpace`, and once with an `IndexSpace`.
Much of the complexity of the `data_specs` come from the requirement that the `source` only indicates _what_ data, and not the format (so that swapping models for others with different input spaces can be done without changing the name of the sources they are requiring), and that different parts in the same experiment can request the same data in different formats at the same time.
It would be good if we find a way to fulfill these requirements in a simpler way.